core/src/main/java/org/apache/iceberg/puffin/PuffinReader.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.puffin;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.RangeReadable;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.puffin.PuffinFormat.Flag;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+import org.apache.iceberg.util.Pair;
+
+public class PuffinReader implements Closeable {
+  // Must not be modified
+  private static final byte[] MAGIC = PuffinFormat.getMagic();
+
+  private final long fileSize;
+  private final SeekableInputStream input;
+  private Integer knownFooterSize;
+  private FileMetadata knownFileMetadata;
+
+  public PuffinReader(InputFile inputFile, @Nullable Long fileSize, @Nullable Long footerSize) {
+    Preconditions.checkNotNull(inputFile, "inputFile is null");
+    this.fileSize = fileSize == null ? inputFile.getLength() : fileSize;
```

Comment:
Seeing as how many of these files are likely going to be small, we might want to consider using `RangeReadable` interface to try to opportunistically read in the whole file with a `readTail` to reduce the number of round-trips to the storage layer.  @findepi this is similar to Trino's `readTail`.

We could just add a default implementation for other FileIO's that mimics the behavior with other calls.

Reply:
@danielcweeks thanks for the suggestion. please see my earlier response https://github.com/apache/iceberg/pull/4537#discussion_r872407417 .  does it answer the concerns here?

Reply:
added "Statistics are informational. A reader can choose to
ignore statistics information. Statistics support is not required to read the table correctly." in https://github.com/apache/iceberg/pull/4945